File: packages/engine/src/notation.ts
/**
 * Notation parsing and serialization.
 *
 * Parsing delegates to cubing.js rather than reimplementing a grammar. That buys the full
 * alg.cubing.net dialect for free — `//` line comments, `F2'`, lowercase wide moves,
 * repeated tokens like `D D`, commutators and conjugates — which is exactly what the
 * reconstruction corpus is written in.
 */
import { Alg, Move as AlgMove } from "cubing/alg";
import { makeMove, type Move } from "./moves.ts";

export class NotationError extends Error {
  override readonly name = "NotationError";
}

/**
 * Parse an algorithm into a flat move list.
 *
 * Commutators, conjugates, and groupings are expanded. Comments, newlines, and pauses are
 * dropped, as are no-op moves such as `R4`.
 *
 * @throws {NotationError} on malformed syntax or an unrecognized move family.
 */
export function parseMoves(text: string): Move[] {
  let alg: Alg;
  try {
    alg = Alg.fromString(text).expand();
  } catch (cause) {
    throw new NotationError(
      `could not parse algorithm: ${(cause as Error).message}`,
      { cause },
    );
  }

  const moves: Move[] = [];
  for (const node of alg.childAlgNodes()) {
    if (!(node instanceof AlgMove)) continue; // comments, newlines, pauses

    // Layer-prefixed moves (`2U`, `3Rw`, `2-3r`) parse with a bare family plus layer
    // numbers, so accepting them blindly would silently apply `U` for `2U`. They are
    // big-cube notation, absent from CFOP reconstructions, and the tables do not model
    // them — reject rather than guess.
    const { innerLayer, outerLayer } = node.quantum;
    if (innerLayer !== null || outerLayer !== null) {
      throw new NotationError(
        `layer-prefixed moves are not supported: ${node.toString()}`,
      );
    }

    const move = makeMove(node.family, node.amount);
    if (move === undefined) {
      throw new NotationError(`unrecognized move family: ${node.toString()}`);
    }
    if (move === null) continue; // whole rotation, e.g. R4
    moves.push(move);
  }
  return moves;
}

/** Serialize one move, e.g. `{ family: "R", amount: -1 }` becomes `R'`. */
export function serializeMove(move: Move): string {
  return `${move.family}${move.amount === 1 ? "" : move.amount === 2 ? "2" : "'"}`;
}

/** Serialize a move list to space-separated notation. */
export function serializeMoves(moves: readonly Move[]): string {
  return moves.map(serializeMove).join(" ");
}

/** Convert a move list to a cubing.js {@link Alg}, for the twisty player or solvers. */
export function toAlg(moves: readonly Move[]): Alg {
  return Alg.fromString(serializeMoves(moves));
}

File: swift/CubingCore/Sources/CubingCore/Notation.swift
/// Parsing and writing move notation.
///
/// The TypeScript delegates parsing to cubing.js, which buys the whole alg.cubing.net dialect for
/// free — commutators, conjugates, groupings — and is the right trade there because the
/// reconstruction corpus is written in it. Swift has no such library, so this is a hand-written
/// parser over the subset the app actually produces and consumes: face turns, wide moves, slices
/// and rotations, with `'` and `2` modifiers, lowercase aliases, and `//` comments.
///
/// What it deliberately does **not** accept is layer-prefixed notation (`2U`, `3Rw`, `2-3r`). The
/// TypeScript rejects those too, and for a reason worth preserving: `2U` is not `u`, the tables do
/// not model it, and accepting it blindly would silently apply `U` instead. Big-cube notation is
/// absent from CFOP reconstructions; rejecting is better than guessing.
///
/// Commutator and conjugate syntax is out of scope here. If the corpus is ever fed to the Swift
/// port directly it will need them, but nothing the app itself writes uses them.

public struct NotationError: Error, CustomStringConvertible {
    public let description: String
    init(_ description: String) { self.description = description }
}

public enum Notation {
    /// Parse an algorithm into a flat move list.
    ///
    /// Comments and no-op moves such as `R4` are dropped; an unknown family or a layer prefix
    /// throws, so a typo fails loudly rather than being applied as something else.
    public static func parse(_ text: String) throws -> [Move] {
        var moves: [Move] = []

        for var token in tokens(of: text) {
            // Layer prefixes: a leading digit, or a range like `2-3`.
            if let first = token.first, first.isNumber || first == "-" {
                throw NotationError("layer-prefixed moves are not supported: \(token)")
            }

            // `<family><digits?><'?>`. The count is parsed as a number rather than matched
            // against "2", so `R4` reads as a whole rotation — a legitimate no-op that appears in
            // reconstructions — instead of failing as an unknown family called "R4".
            var inverted = false
            while let last = token.last, last == "'" || last == "\u{2019}" {
                inverted = true
                token.removeLast()
            }
            var digits = ""
            while let last = token.last, last.isNumber {
                digits.insert(last, at: digits.startIndex)
                token.removeLast()
            }
            var amount = digits.isEmpty ? 1 : (Int(digits) ?? 1)
            if inverted { amount = -amount }

            guard !token.isEmpty else { throw NotationError("empty move") }
            guard let move = MoveTables.make(family: token, amount: amount) else {
                // `make` returns nil for an unknown family *and* for a whole rotation such as
                // `R4`. Only the first is an error, so resolve the family to tell them apart.
                if MoveTables.resolve(token) != nil { continue }
                throw NotationError("unrecognized move family: \(token)")
            }
            moves.append(move)
        }
        return moves
    }

    /// Whitespace-separated tokens, with `//` comments to end of line removed.
    ///
    /// Scanned by hand rather than with `range(of:)` so this module needs no Foundation import —
    /// the core is pure computation and is worth keeping that way.
    private static func tokens(of text: String) -> [String] {
        text
            .split(separator: "\n", omittingEmptySubsequences: false)
            .map(dropComment)
            .flatMap { $0.split(whereSeparator: { $0 == " " || $0 == "\t" || $0 == "\r" }) }
            .map(String.init)
            .filter { !$0.isEmpty }
    }

    private static func dropComment(_ line: Substring) -> Substring {
        var previous: Character?
        for index in line.indices {
            if line[index] == "/", previous == "/" {
                return line[line.startIndex..<line.index(before: index)]
            }
            previous = line[index]
        }
        return line
    }

    public static func write(_ moves: [Move]) -> String {
        moves.map(\.notation).joined(separator: " ")
    }
}
